File: rust/src/r1cs.rs
use crate::commitment::CommitmentScheme;
use crate::field::{field_zeros, Field};
use crate::matrix::MatrixOps;

/// R1CS constraint system: Az . Bz = Cz.
pub struct R1CSShape<F: Field, M: MatrixOps<F>> {
    pub a: M,
    pub b: M,
    pub c: M,
    pub m: usize,
    pub n: usize,
    pub l: usize,
    _phantom: std::marker::PhantomData<F>,
}

impl<F: Field, M: MatrixOps<F>> R1CSShape<F, M> {
    pub fn new(a: M, b: M, c: M, m: usize, n: usize, l: usize) -> Self {
        R1CSShape {
            a,
            b,
            c,
            m,
            n,
            l,
            _phantom: std::marker::PhantomData,
        }
    }

    /// Check Az . Bz = Cz where z = (1, x, w).
    pub fn is_satisfied(&self, x: &[F], w: &[F]) -> bool {
        let z = self.make_z(x, w);
        let az = self.a.mul_vec(&z);
        let bz = self.b.mul_vec(&z);
        let cz = self.c.mul_vec(&z);
        az.iter()
            .zip(bz.iter())
            .zip(cz.iter())
            .all(|((&a, &b), &c)| a.mul(b) == c)
    }

    /// Check Az . Bz = u*Cz + E where z = (u, x, W).
    pub fn is_relaxed_satisfied<C: CommitmentScheme<F>>(
        &self,
        instance: &RelaxedR1CSInstance<F, C>,
        witness: &RelaxedR1CSWitness<F>,
    ) -> bool {
        let z = self.make_relaxed_z(instance.u, &instance.x, &witness.w);
        let az = self.a.mul_vec(&z);
        let bz = self.b.mul_vec(&z);
        let cz = self.c.mul_vec(&z);
        az.iter()
            .zip(bz.iter())
            .zip(cz.iter())
            .zip(witness.e.iter())
            .all(|(((&a, &b), &c), &e)| a.mul(b) == instance.u.mul(c).add(e))
    }

    /// Build z = (1, x, w).
    pub fn make_z(&self, x: &[F], w: &[F]) -> Vec<F> {
        let mut z = Vec::with_capacity(self.n);
        z.push(F::one());
        z.extend_from_slice(x);
        z.extend_from_slice(w);
        z
    }

    /// Build z = (u, x, w).
    pub fn make_relaxed_z(&self, u: F, x: &[F], w: &[F]) -> Vec<F> {
        let mut z = Vec::with_capacity(self.n);
        z.push(u);
        z.extend_from_slice(x);
        z.extend_from_slice(w);
        z
    }
}

/// Public part of a Relaxed R1CS instance.
pub struct RelaxedR1CSInstance<F: Field, C: CommitmentScheme<F>> {
    pub com_e: C,
    pub u: F,
    pub x: Vec<F>,
    pub com_w: C,
}

/// Private part of a Relaxed R1CS instance.
pub struct RelaxedR1CSWitness<F: Field> {
    pub e: Vec<F>,
    pub w: Vec<F>,
}

/// Lift a standard R1CS instance (x, w) to Relaxed R1CS with u=1, E=0.
pub fn r1cs_to_relaxed<F: Field, M: MatrixOps<F>, C: CommitmentScheme<F>>(
    shape: &R1CSShape<F, M>,
    x: &[F],
    w: &[F],
) -> (RelaxedR1CSInstance<F, C>, RelaxedR1CSWitness<F>) {
    let e = field_zeros::<F>(shape.m);
    let com_e = C::commit(&e);
    let com_w = C::commit(w);
    let instance = RelaxedR1CSInstance {
        com_e,
        u: F::one(),
        x: x.to_vec(),
        com_w,
    };
    let witness = RelaxedR1CSWitness {
        e,
        w: w.to_vec(),
    };
    (instance, witness)
}

/// Create a trivial relaxed instance (u=0, E=0, W=0, x=0).
pub fn trivial_relaxed<F: Field, M: MatrixOps<F>, C: CommitmentScheme<F>>(
    shape: &R1CSShape<F, M>,
) -> (RelaxedR1CSInstance<F, C>, RelaxedR1CSWitness<F>) {
    let e = field_zeros::<F>(shape.m);
    let w = field_zeros::<F>(shape.n - 1 - shape.l);
    let x = field_zeros::<F>(shape.l);
    let com_e = C::commit(&e);
    let com_w = C::commit(&w);
    let instance = RelaxedR1CSInstance {
        com_e,
        u: F::zero(),
        x,
        com_w,
    };
    let witness = RelaxedR1CSWitness { e, w };
    (instance, witness)
}

File: nano_nova/r1cs.py
"""
R1CS and Relaxed R1CS data structures.

An R1CS instance encodes a computation as:
    Az ∘ Bz = Cz
where z = (1, x, w), A/B/C are sparse matrices, and ∘ is Hadamard product.

A Relaxed R1CS generalizes this to:
    Az ∘ Bz = u·Cz + E
where u is a scalar and E is an error vector. For a "real" instance, u=1 and E=0.

Nova paper reference: Section 4 (Folding Scheme for R1CS)
"""

from dataclasses import dataclass

import numpy as np

from .field import GF, field_zeros


@dataclass
class R1CSShape:
    """The structure of an R1CS — matrices A, B, C and dimensions.

    Attributes:
        A, B, C: Constraint matrices over GF(p). Shape: (m, n) where
                  m = number of constraints, n = 1 + |x| + |w|.
        m: Number of constraints.
        n: Total witness length (1 + public inputs + private witness).
        l: Number of public inputs.
    """

    A: np.ndarray  # GF array, shape (m, n)
    B: np.ndarray  # GF array, shape (m, n)
    C: np.ndarray  # GF array, shape (m, n)
    m: int  # number of constraints
    n: int  # total variables (1 + l + witness_size)
    l: int  # number of public inputs

    def is_satisfied(self, x, w) -> bool:
        """Check if (x, w) satisfies Az ∘ Bz = Cz."""
        z = self._make_z(x, w)
        Az = self.A @ z
        Bz = self.B @ z
        Cz = self.C @ z
        return np.all(Az * Bz == Cz)

    def is_relaxed_satisfied(
        self, instance: "RelaxedR1CSInstance", witness: "RelaxedR1CSWitness"
    ) -> bool:
        """Check if a relaxed instance satisfies Az ∘ Bz = u·Cz + E.

        For relaxed R1CS, z = (u, x, W) — the first element is u, not 1.
        This is because folding combines z' = z₁ + r·z₂, making the first
        element u₁ + r·u₂ = u'. The matrices select u from position 0.
        """
        z = self._make_relaxed_z(instance.u, instance.x, witness.W)
        Az = self.A @ z
        Bz = self.B @ z
        Cz = self.C @ z
        lhs = Az * Bz
        rhs = instance.u * Cz + witness.E
        return np.all(lhs == rhs)

    def _make_z(self, x, w):
        """Construct z = (1, x, w) for standard R1CS."""
        one = GF(np.array([1]))
        return GF(np.concatenate([one, x, w]))

    def _make_relaxed_z(self, u, x, w):
        """Construct z = (u, x, w) for relaxed R1CS."""
        u_arr = GF(np.array([int(u)]))
        return GF(np.concatenate([u_arr, x, w]))


@dataclass
class RelaxedR1CSInstance:
    """Public part of a Relaxed R1CS instance.

    Nova paper: Definition 7 — a relaxed instance is (com_E, u, x, com_W).
    We store commitments as opaque values and keep u, x as field elements.

    Attributes:
        com_E: Commitment to the error vector E.
        u: Relaxation scalar. u=1 for "real" instances, u=0 for trivial.
        x: Public input/output vector.
        com_W: Commitment to the witness W.
    """

    com_E: object  # commitment (opaque)
    u: object  # field element
    x: np.ndarray  # GF array of public inputs
    com_W: object  # commitment (opaque)


@dataclass
class RelaxedR1CSWitness:
    """Private part of a Relaxed R1CS instance.

    Attributes:
        E: Error vector. E=0 for real instances.
        W: Witness vector.
    """

    E: np.ndarray  # GF array, length m
    W: np.ndarray  # GF array, length (n - 1 - l)


def r1cs_to_relaxed(
    shape: R1CSShape, x, w, commit_fn
) -> tuple[RelaxedR1CSInstance, RelaxedR1CSWitness]:
    """Lift a standard R1CS instance to a Relaxed R1CS instance.

    A satisfied R1CS instance (x, w) becomes a relaxed instance with u=1, E=0.

    Args:
        shape: The R1CS structure.
        x: Public inputs.
        w: Witness.
        commit_fn: Commitment function Com(vector) -> commitment.

    Returns:
        (instance, witness) pair for the relaxed system.
    """
    E = field_zeros(shape.m)
    u = GF(1)
    com_E = commit_fn(E)
    com_W = commit_fn(w)
    instance = RelaxedR1CSInstance(com_E=com_E, u=u, x=x, com_W=com_W)
    witness = RelaxedR1CSWitness(E=E, W=w)
    return instance, witness


def trivial_relaxed(shape: R1CSShape, commit_fn) -> tuple[RelaxedR1CSInstance, RelaxedR1CSWitness]:
    """Create a trivial relaxed instance (u=0, E=0, W=0, x=0).

    This is the starting accumulator for IVC. With u=0, W=0, x=0, E=0, the
    relaxed relation becomes Az∘Bz = 0·Cz + 0. Since z = (u, x, W) = (0, 0, 0),
    every matrix product Az, Bz, Cz is zero, so 0 = 0 holds trivially.

    Args:
        shape: The R1CS structure.
        commit_fn: Commitment function.

    Returns:
        Trivial (instance, witness) pair.
    """
    E = field_zeros(shape.m)
    W = field_zeros(shape.n - 1 - shape.l)
    x = field_zeros(shape.l)
    u = GF(0)
    com_E = commit_fn(E)
    com_W = commit_fn(W)
    instance = RelaxedR1CSInstance(com_E=com_E, u=u, x=x, com_W=com_W)
    witness = RelaxedR1CSWitness(E=E, W=W)
    return instance, witness
